File: Scala/src/com/yourtion/Pattern16/CompositionExamples.scala
package com.yourtion.Pattern16

/**
  * Created by Yourtion on 05/04/2017.
  */
object CompositionExamples {

  val appendA = (s: String) => s + "a"
  val appendB = (s: String) => s + "b"
  val appendC = (s: String) => s + "c"

  case class HttpRequest(
                          headers: Map[String, String],
                          payload: String,
                          principal: Option[String] = None)

  def checkAuthorization(request: HttpRequest) = {
    val authHeader = request.headers.get("Authorization")
    val mockPrincipal = authHeader match {
      case Some(headerValue) => Some("AUser")
      case _ => None
    }
    request.copy(principal = mockPrincipal)
  }
  def logFingerprint(request: HttpRequest) = {
    val fingerprint = request.headers.getOrElse("X-RequestFingerprint", "")
    println("FINGERPRINT=" + fingerprint)
    request
  }

  def composeFilters(filters: Seq[Function1[HttpRequest, HttpRequest]]) =
    filters.reduce {
      (allFilters, currentFilter) => allFilters compose currentFilter
    }

  def run(): Unit = {
    println("CompositionExamples :")
    val appendCBA = appendA compose appendB compose appendC
    println("appendA + appendB + appendC: " + appendCBA("z"))
    println()

    val filterChain = composeFilters(Vector(checkAuthorization, logFingerprint))
    val requestHeader = Map("Authorization" -> "Auth", "X-RequestFingerprint" -> "fingerprint")
    val request = HttpRequest(requestHeader, "body")
    println("FilterChain: " + filterChain(request))
  }
}

File: Clojure/src/com/yourtion/Pattern16/composition_examples.clj
(ns com.yourtion.Pattern16.composition-examples)

(defn append-a [s] (str s "a"))
(defn append-b [s] (str s "b"))
(defn append-c [s] (str s "c"))

(def request
  {:headers
         {"Authorization" "auth"
          "X-RequestFingerprint" "fingerprint"}
   :body "body"})

(defn check-authorization [request]
  (let [auth-header (get-in request [:headers "Authorization"])]
    (assoc
      request
      :principal
      (if-not (nil? auth-header)
        "AUser"))))

(defn log-fingerprint [request]
  (let [fingerprint (get-in request [:headers "X-RequestFingerprint"])]
    (println (str "FINGERPRINT=" fingerprint))
    request))

(defn compose-filters [filters]
  (reduce
    (fn [all-filters, current-filter] (comp all-filters current-filter))
    filters))

(defn run []
  (println "CompositionExamples: ")
  (def append-cba (comp append-a append-b append-c))
  (println (append-cba "z"))
  (println "")

  (def request {:headers
                {"Authorization" "auth"
                 "X-RequestFingerprint" "fingerprint"}
                :body "Body"})
  (def filter-chain (compose-filters [check-authorization log-fingerprint]))
  (println (filter-chain request))
  (println ""))
